File: source/includes/write/replace.kt
import com.mongodb.client.model.Filters
import com.mongodb.client.model.ReplaceOptions
import com.mongodb.kotlin.client.MongoClient

// start-data-class
data class Restaurant(
    val name: String,
    val borough: String,
    val cuisine: String,
    val owner: String?,
)
// end-data-class

fun main() {
    val uri = "<connection string>"

    val mongoClient = MongoClient.create(uri)
    val database = mongoClient.getDatabase("sample_restaurants")
    val collection = database.getCollection<Restaurant>("restaurants")

    // start-replace-one
    val filter = Filters.eq(Restaurant::name.name, "Primola Restaurant")
    val replacement = Restaurant(
        "Frutti Di Mare",
        "Queens",
        "Seafood",
        owner = "Sal Thomas"
    )
    val result = collection.replaceOne(filter, replacement)
    // end-replace-one

    // start-replace-options
    val opts = ReplaceOptions().upsert(true)
    val result = collection.replaceOne(filter, replacement, opts)
    // end-replace-options

}

File: upcoming/source/includes/write/replace.scala
import org.mongodb.scala._
import org.mongodb.scala.bson.Document
import org.mongodb.scala.model.Filters.equal
import org.mongodb.scala.model.ReplaceOptions
import org.mongodb.scala.result.UpdateResult

object Replace {

  def main(args: Array[String]): Unit = {
    val mongoClient = MongoClient("<connection string URI>")

    // start-db-coll
    val database: MongoDatabase = mongoClient.getDatabase("sample_restaurants")
    val collection: MongoCollection[Document] = database.getCollection("restaurants")
    // end-db-coll

    // start-replace-one
    val filter = equal("name", "Primola Restaurant")
    val replacement = Document(
      "name" -> "Frutti Di Mare",
      "borough" -> "Queens",
      "cuisine" -> "Seafood",
      "owner" -> "Sal Thomas"
    )
    val observable: Observable[UpdateResult] = collection.replaceOne(filter, replacement)

    observable.subscribe(new Observer[UpdateResult] {
      override def onNext(result: UpdateResult): Unit = println(s"Replaced document count: ${result.getModifiedCount}")
      override def onError(e: Throwable): Unit = println(s"Failed: ${e.getMessage}")
      override def onComplete(): Unit = println("Completed")
    })
    // end-replace-one

    // start-replace-options
    val options = ReplaceOptions().upsert(true)
    val observable: Observable[UpdateResult] = collection.replaceOne(filter, replacement, options)

    observable.subscribe(new Observer[UpdateResult] {
      override def onNext(result: UpdateResult): Unit = println(s"Replaced document count: ${result.getModifiedCount}")
      override def onError(e: Throwable): Unit = println(s"Failed: ${e.getMessage}")
      override def onComplete(): Unit = println("Completed")
    })
    // end-replace-options

    Thread.sleep(1000)
    mongoClient.close()
  }
}
